## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#248571c1-a237-4d2f-99d7-cfe7f00c162d) |
| mint-gateway-azure.sh | :x: [more...](#f28970c2-2e2c-4ced-a450-46af9fee69d4) |
| mint-gateway-nas.sh | :x: [more...](#3eba3f4f-358c-4257-b986-dd9f95e5b022) |
<a name="248571c1-a237-4d2f-99d7-cfe7f00c162d"></a>
### 10412-94c42bb/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30976
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp abc4bda14c5b:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 2 minutes and 21 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 1 hours 55 minutes and 8 seconds
{
  "name": "mc",
  "duration": "6873839",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-26978 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-7052  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-7052/datafile-0-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-7052/datafile-1-kB`\n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-7052/datafile-1-MB`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-7052/datafile-1-b`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-7052/datafile-10-MB`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-7052/datafile-1.03-MB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-7052/datafile-100-kB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-7052/datafile-10-kB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-7052/datafile-129-MB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-7052/datafile-11-MB`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-7052/datafile-5-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-7052/datafile-33-kB`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-7052/datafile-6-MB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-7052/datafile-5243880-b`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-7052/datafile-65-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. A timeout occurred while trying to lock a resource, please reduce your request rate\nTotal: 0 B, Transferred: 430.38 MiB, Speed: 64.12 KiB/s"
}
(7/15) Running minio-dotnet tests ... done in 40 seconds
(8/15) Running minio-go tests ... done in 1 minutes and 24 seconds
(9/15) Running minio-java tests ... done in 1 minutes and 13 seconds
(10/15) Running minio-js tests ... done in 49 seconds
(11/15) Running minio-py tests ... done in 3 minutes and 15 seconds
(12/15) Running s3cmd tests ... done in 20 seconds
(13/15) Running s3select tests ... done in 8 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="3eba3f4f-358c-4257-b986-dd9f95e5b022"></a>
### 10412-94c42bb/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30833
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0eaccd3f00fe:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 3 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 16 minutes and 1 seconds
{
  "name": "mc",
  "duration": "951836",
  "function": "test_mirror_list_objects",
  "status": "FAIL",
  "error": "mirror and list differs"
}
(7/15) Running minio-dotnet tests ... done in 33 seconds
(8/15) Running minio-go tests ... done in 33 seconds
(9/15) Running minio-java tests ... done in 11 seconds
(10/15) Running minio-js tests ... done in 30 seconds
(11/15) Running minio-py tests ... done in 59 seconds
(12/15) Running s3cmd tests ... done in 12 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="f28970c2-2e2c-4ced-a450-46af9fee69d4"></a>
### 10412-94c42bb/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30640
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 22a93d13df1f:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 6 minutes and 15 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... FAILED in 42 seconds
{
  "name": "awscli",
  "duration": 3590,
  "function": "aws --endpoint-url http://minio-c2.minio.io:30640 s3api copy-object --bucket awscli-mint-test-bucket-15741 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-15741/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected a861ac626e0bee4421a50737c581233c, got 0x8D8504C9E5FD811-1"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 3 minutes and 33 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 37 seconds
(8/15) Running minio-go tests ... done in 5 minutes and 49 seconds
(9/15) Running minio-java tests ... FAILED in 7 minutes and 47 seconds
{
  "name": "minio-java",
  "function": "selectObjectContent(SelectObjectContentArgs args)",
  "args": "sqlExpression: select * from S3Object, requestProgress: true",
  "duration": 493,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InternalError, message = We encountered an internal error, please try again., bucketName = minio-java-test-2gj30dc, objectName = minio-java-test-33q9m75, resource = /minio-java-test-2gj30dc/minio-java-test-33q9m75, requestId = 163162DF7BCDD470, hostId = a3c70b08-ae15-4aa2-a7e5-2b1659f12c2e)\nrequest={method=POST, url=http://minio-c2.minio.io:30640/minio-java-test-2gj30dc/minio-java-test-33q9m75?select=&select-type=2, headers=Host: minio-c2.minio.io:30640\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/7.1.0\nContent-MD5: 8xo/e3irSozlXjiX3z0KMg==\nx-amz-content-sha256: 753534621e04bc3c95aec4d971a9acb8d059186d08e491ab902bbeca6b26800b\nx-amz-date: 20200903T212054Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200903/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=500, headers=Accept-Ranges: bytes\nContent-Length: 393\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.2020-09-03T20-45-00Z\nVary: Origin\nX-Amz-Request-Id: 163162DF7BCDD470\nX-Xss-Protection: 1; mode=block\nDate: Thu, 03 Sep 2020 21:20:54 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1135), io.minio.MinioClient.execute(MinioClient.java:935), io.minio.MinioClient.executePost(MinioClient.java:1239), io.minio.MinioClient.selectObjectContent(MinioClient.java:6412), FunctionalTest.selectObjectContent_test1(FunctionalTest.java:3397), FunctionalTest.runTests(FunctionalTest.java:3835), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(9/15) Running minio-js tests ... done in 2 minutes and 33 seconds
(10/15) Running minio-py tests ... done in 13 minutes and 7 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 11 seconds
(12/15) Running s3select tests ... FAILED in 18 seconds
{
  "name": "s3select:test_csv_output_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-db9d92db-b4db-4fe8-a7c0-e896921bc8db"
  },
  "duration": 5417,
  "message": "Test test_8 unexpectedly failed with: 'event-type'",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/s3select/csv.py\", line 39, in test_sql_api\n    for d in data.stream(10*1024):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/select/reader.py\", line 211, in stream\n    message = self.__extract_message()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/select/reader.py\", line 161, in __extract_message\n    event_type = header_map[\"event-type\"]\nKeyError: 'event-type'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 56, in main\n    test_csv_output_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 180, in test_csv_output_custom_quote_char\n    input_data, sql_opts, expected_output)\n  File \"/mint/run/core/s3select/csv.py\", line 45, in test_sql_api\n    'Test {} unexpectedly failed with: {}'.format(test_name, select_err))\nValueError: Test test_8 unexpectedly failed with: 'event-type'\n",
  "status": "FAIL"
}
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10412-94c42bb

--- a/cmd/api-errors.go
+++ b/cmd/api-errors.go
@@ -109,6 +109,7 @@ const (
 	ErrReplicationDestinationNotFoundError
 	ErrReplicationDestinationMissingLock
 	ErrReplicationTargetNotFoundError
+	ErrReplicationRemoteConnectionError
 	ErrBucketRemoteIdenticalToSource
 	ErrBucketRemoteAlreadyExists
 	ErrBucketRemoteArnTypeInvalid
@@ -821,6 +822,11 @@ var errorCodes = errorCodeMap{
 	ErrReplicationTargetNotFoundError: {
 		Code:           "XminioAdminReplicationTargetNotFoundError",
 		Description:    "The replication target does not exist",
+		HTTPStatusCode: http.StatusNotFound,
+	},
+	ErrReplicationRemoteConnectionError: {
+		Code:           "XminioAdminReplicationRemoteConnectionError",
+		Description:    "Remote service endpoint or target bucket not available",
 		HTTPStatusCode: http.StatusNotFound,
 	},
 	ErrBucketRemoteIdenticalToSource: {
@@ -1906,6 +1912,8 @@ func toAPIErrorCode(ctx context.Context, err error) (apiErr APIErrorCode) {
 		apiErr = ErrReplicationDestinationMissingLock
 	case BucketRemoteTargetNotFound:
 		apiErr = ErrReplicationTargetNotFoundError
+	case BucketRemoteConnectionErr:
+		apiErr = ErrReplicationRemoteConnectionError
 	case BucketRemoteAlreadyExists:
 		apiErr = ErrBucketRemoteAlreadyExists
 	case BucketRemoteArnTypeInvalid:

--- a/cmd/bucket-targets.go
+++ b/cmd/bucket-targets.go
@@ -21,6 +21,7 @@ import (
 	"net/http"
 	"sync"
 
+	minio "github.com/minio/minio-go/v7"
 	miniogo "github.com/minio/minio-go/v7"
 	"github.com/minio/minio-go/v7/pkg/credentials"
 	"github.com/minio/minio/pkg/bucket/versioning"
@@ -91,13 +92,13 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
 		}
 		vcfg, err := clnt.GetBucketVersioning(ctx, tgt.TargetBucket)
 		if err != nil {
-			if isErrBucketNotFound(err) {
+			if minio.ToErrorResponse(err).Code == "NoSuchBucket" {
 				return BucketRemoteTargetNotFound{Bucket: tgt.TargetBucket}
 			}
-			if vcfg.Status != string(versioning.Enabled) {
-				return BucketReplicationTargetNotVersioned{Bucket: tgt.TargetBucket}
-			}
-			return err
+			return BucketRemoteConnectionErr{Bucket: tgt.TargetBucket}
+		}
+		if vcfg.Status != string(versioning.Enabled) {
+			return BucketReplicationTargetNotVersioned{Bucket: tgt.TargetBucket}
 		}
 	}
 

--- a/cmd/object-api-errors.go
+++ b/cmd/object-api-errors.go
@@ -376,6 +376,13 @@ func (e BucketRemoteTargetNotFound) Error() string {
 	return "Remote target not found: " + e.Bucket
 }
 
+// BucketRemoteConnectionErr remote target connection failure.
+type BucketRemoteConnectionErr GenericError
+
+func (e BucketRemoteConnectionErr) Error() string {
+	return "Remote service endpoint or target bucket not available: " + e.Bucket
+}
+
 // BucketRemoteAlreadyExists remote already exists for this target type.
 type BucketRemoteAlreadyExists GenericError
 

